File: swift/Sources/benchmarks/Matmul.swift
import Dispatch
import Foundation

class MatmulBase: BenchmarkProtocol {
  private var _n: Int64 = 0
  private var _resultVal: UInt32 = 0
  private var _a: [[Double]] = []
  private var _b: [[Double]] = []
  private let lock = NSLock()

  var n: Int64 {
    return _n
  }

  var resultVal: UInt32 {
    get { return _resultVal }
    set { _resultVal = newValue }
  }

  var a: [[Double]] {
    return _a
  }

  var b: [[Double]] {
    return _b
  }

  func prepare() {
    _n = configValue("n") ?? 0
    _a = matgen(Int(_n))
    _b = matgen(Int(_n))
    _resultVal = 0
  }

  func matgen(_ n: Int) -> [[Double]] {
    let tmp = 1.0 / Double(n) / Double(n)
    var a = [[Double]](repeating: [Double](repeating: 0, count: n), count: n)

    for i in 0..<n {
      for j in 0..<n {
        a[i][j] = tmp * Double(i - j) * Double(i + j)
      }
    }
    return a
  }

  func transpose(_ b: [[Double]]) -> [[Double]] {
    let n = b.count
    var bT = [[Double]](repeating: [Double](repeating: 0, count: n), count: n)

    for i in 0..<n {
      for j in 0..<n {
        bT[j][i] = b[i][j]
      }
    }
    return bT
  }

  func matmulSequential(_ a: [[Double]], _ b: [[Double]]) -> [[Double]] {
    let n = a.count
    let bT = transpose(b)
    var c = [[Double]](repeating: [Double](repeating: 0, count: n), count: n)

    for i in 0..<n {
      let ai = a[i]
      var ci = c[i]
      for j in 0..<n {
        var s = 0.0
        let bTj = bT[j]

        for k in 0..<n {
          s += ai[k] * bTj[k]
        }
        ci[j] = s
      }
      c[i] = ci
    }
    return c
  }

  func matmulParallel(_ a: [[Double]], _ b: [[Double]], numThreads: Int) -> [[Double]] {
    let n = a.count
    let bT = transpose(b)
    var c = [[Double]](repeating: [Double](repeating: 0, count: n), count: n)

    DispatchQueue.concurrentPerform(iterations: numThreads) { threadId in
      let startRow = threadId
      for i in stride(from: startRow, to: n, by: numThreads) {
        let ai = a[i]
        var row = [Double](repeating: 0.0, count: n)

        for j in 0..<n {
          var sum = 0.0
          let bTj = bT[j]

          for k in 0..<n {
            sum += ai[k] * bTj[k]
          }
          row[j] = sum
        }

        lock.lock()
        c[i] = row
        lock.unlock()
      }
    }

    return c
  }

  func getNumThreads() -> Int {
    return 1
  }

  func run(iterationId: Int) {

  }

  var checksum: UInt32 {
    return _resultVal
  }

  func name() -> String {
    return "MatmulBase"
  }
}

final class Matmul1T: MatmulBase {
  override func name() -> String {
    return "Matmul::Single"
  }

  override func run(iterationId: Int) {
    let c = matmulSequential(a, b)
    let center = c[Int(n) / 2][Int(n) / 2]
    resultVal &+= Helper.checksumF64(center)
  }
}

class MatmulParallelBase: MatmulBase {
  override func getNumThreads() -> Int {
    return 4
  }

  override func run(iterationId: Int) {
    let c = matmulParallel(a, b, numThreads: getNumThreads())
    let center = c[Int(n) / 2][Int(n) / 2]
    resultVal &+= Helper.checksumF64(center)
  }
}

final class Matmul4T: MatmulParallelBase {
  override func name() -> String {
    return "Matmul::T4"
  }

  override func getNumThreads() -> Int {
    return 4
  }
}

final class Matmul8T: MatmulParallelBase {
  override func name() -> String {
    return "Matmul::T8"
  }

  override func getNumThreads() -> Int {
    return 8
  }
}

final class Matmul16T: MatmulParallelBase {
  override func name() -> String {
    return "Matmul::T16"
  }

  override func getNumThreads() -> Int {
    return 16
  }
}

File: golang/matmul.go
package main

import (
	"runtime"
	"sync"
)

func matgen(n int) [][]float64 {
	tmp := 1.0 / float64(n) / float64(n)
	a := make([][]float64, n)
	for i := range a {
		a[i] = make([]float64, n)
		for j := range a[i] {
			a[i][j] = tmp * float64(i-j) * float64(i+j)
		}
	}
	return a
}

func transpose(b [][]float64) [][]float64 {
	n := len(b)
	bT := make([][]float64, n)
	for i := range bT {
		bT[i] = make([]float64, n)
		for j := 0; j < n; j++ {
			bT[i][j] = b[j][i]
		}
	}
	return bT
}

func matmulSequential(a, b [][]float64) [][]float64 {
	n := len(a)
	bT := transpose(b)

	c := make([][]float64, n)
	for i := range c {
		c[i] = make([]float64, n)
		ai := a[i]
		ci := c[i]

		for j := 0; j < n; j++ {
			s := 0.0
			bTj := bT[j]

			for k := 0; k < n; k++ {
				s += ai[k] * bTj[k]
			}
			ci[j] = s
		}
	}
	return c
}

func matmulParallel(a, b [][]float64, numThreads int) [][]float64 {
	n := len(a)
	bT := transpose(b)

	c := make([][]float64, n)
	for i := range c {
		c[i] = make([]float64, n)
	}

	runtime.GOMAXPROCS(numThreads)

	var wg sync.WaitGroup
	rowsPerThread := (n + numThreads - 1) / numThreads

	for w := 0; w < numThreads; w++ {
		wg.Add(1)
		go func(threadID int) {
			defer wg.Done()
			startRow := threadID * rowsPerThread
			endRow := startRow + rowsPerThread
			if endRow > n || threadID == numThreads-1 {
				endRow = n
			}

			for i := startRow; i < endRow; i++ {
				ai := a[i]
				ci := c[i]
				for j := 0; j < n; j++ {
					sum := 0.0
					bTj := bT[j]

					for k := 0; k < n; k++ {
						sum += ai[k] * bTj[k]
					}
					ci[j] = sum
				}
			}
		}(w)
	}

	wg.Wait()
	return c
}

type BaseMatmul struct {
	BaseBenchmark
	n      int64
	result uint32
	a      [][]float64
	b      [][]float64
}

func (m *BaseMatmul) Prepare() {
	m.n = m.ConfigVal("n")
	n := int(m.n)
	m.a = matgen(n)
	m.b = matgen(n)
}

type Matmul1T struct {
	BaseMatmul
}

func (m *Matmul1T) Run(iteration_id int) {
	c := matmulSequential(m.a, m.b)
	m.result += ChecksumFloat64(c[m.n>>1][m.n>>1])
}

func (m *Matmul1T) Checksum() uint32 {
	return m.result
}

func (m *Matmul1T) name() string {
	return "Matmul::Single"
}

type Matmul4T struct {
	BaseMatmul
}

func (m *Matmul4T) Run(iteration_id int) {
	c := matmulParallel(m.a, m.b, 4)
	m.result += ChecksumFloat64(c[m.n>>1][m.n>>1])
}

func (m *Matmul4T) Checksum() uint32 {
	return m.result
}

func (m *Matmul4T) name() string {
	return "Matmul::T4"
}

type Matmul8T struct {
	BaseMatmul
}

func (m *Matmul8T) Run(iteration_id int) {
	c := matmulParallel(m.a, m.b, 8)
	m.result += ChecksumFloat64(c[m.n>>1][m.n>>1])
}

func (m *Matmul8T) Checksum() uint32 {
	return m.result
}

func (m *Matmul8T) name() string {
	return "Matmul::T8"
}

type Matmul16T struct {
	BaseMatmul
}

func (m *Matmul16T) Run(iteration_id int) {
	c := matmulParallel(m.a, m.b, 16)
	m.result += ChecksumFloat64(c[m.n>>1][m.n>>1])
}

func (m *Matmul16T) Checksum() uint32 {
	return m.result
}

func (m *Matmul16T) name() string {
	return "Matmul::T16"
}
